_drafts/advanced-gdb.md

```diff
@@ -0,0 +1,993 @@
+
+I'm looking at a Zephyr RTOS based system and one of the stack regions is called `my_stack_area`. Let's dump the entire contents.
+
+First we find the size of the stack:
+
+```
+(gdb) p sizeof(my_stack_area)
+$1 = 2980
+```
+
+It's 2980 bytes, so we want to print 2980/4 = 745 words. That should be `x/745a` then.
+
+```
+(gdb) x/745a my_stack_area
+0x2000a3c0 <my_stack_area>:	0xaaaaaaaa	0xaaaaaaaa	0xaaaaaaaa	0xaaaaaaaa
+0x2000a3d0 <my_stack_area+16>:	0xaaaaaaaa	0xaaaaaaaa	0xaaaaaaaa	0xaaaaaaaa
+... A lot more 0xaaaaaaaa
+0x2000a8d0 <my_stack_area+1296>:	0xaaaaaaaa	0xaaaaaaaa	0x190	0x2000a900 <my_stack_area+1344>
+0x2000a8e0 <my_stack_area+1312>:	0x18f	0x0	0x2000a298 <my_stack_area2+152>	0x18f
+0x2000a8f0 <my_stack_area+1328>:	0x20000434 <my_work_q>	0x8021063 <process_accel_data_worker_task>	0x0	0x802108d <process_accel_data_worker_task+42>
+0x2000a900 <my_stack_area+1344>:	0x1	0x20002	0x40002	0x60002
+0x2000a910 <my_stack_area+1360>:	0x80002	0xc0004	0x100004	0x160006
+0x2000a920 <my_stack_area+1376>:	0x1c0006	0x240008	0x2c0008	0x38000c
+... A lot of accel int32_t readings
+0x2000af30 <my_stack_area+2928>:	0x92f5648a	0xf83c6547	0x5d9a654f	0xc39f6605
+0x2000af40 <my_stack_area+2944>:	0x0	0x8021505 <z_work_q_main+68>	0x0	0x80214c1 <z_work_q_main>
+0x2000af50 <my_stack_area+2960>:	0x0	0x80214b9 <z_thread_entry+12>	0x0	0xaaaaaaaa
+0x2000af60 <my_stack_area+2976>:	0xaaaaaaaa
+```
+
+We can easily see some references to functions in this stack, such as `process_accel_data_worker_task`, `z_work_q_main`, and `z_thread_entry`. This stack dump technique is especially useful if your GDB provides you with no backtrace information, such as:
+
+```
+(gdb) bt
+#0  0x00015f5a in ?? ()
+```
+
+[Reference](https://sourceware.org/gdb/current/onlinedocs/gdb/Memory.html)
+
+{:.no_toc}
+
+### Searching Memory using `find`
+
+Sometimes you know a pattern that you are looking for in memory, and you want to quickly find out if it exists in memory on the system. Maybe it's a magic string or a specific 4-byte pattern, like `0xdeadbeef`.
+
+Let's search for the string `shell_uart`, which is the task name of a thread in my Zephyr system. I'll search the entire writeable RAM space, which can be found by running the `info files` command mentioned previously.
+
+```
+(gdb) find 0x20000000, 0x200117e0, "shell_uart"
+0x2000121c <shell_uart_thread+104>
+1 pattern found.
+```
+
+If we use `x/s` to examine that memory, we can see that it is indeed `shell_uart`.
+
+```
+(gdb) x/s 0x2000121c
+0x2000121c <shell_uart_thread+104>:	"shell_uart"
+```
+
+The `find` command can also be useful for finding pointers pointing to arbitrary structs. For example, I want to find all the pointers that contain a reference to the variable `mgmt_thread_data`.
+
+```
+(gdb) find 0x20000000, 0x200117e0, &mgmt_thread_data
```

Comment:
good example!

Reply:
Thanks! It'd be neat to have something use that to find dangling references to free'd heap allocations.